internal/auth/providers/okta.go

```diff
@@ -0,0 +1,416 @@
+}
+
+func (p *OktaProvider) cbStateChange(from, to circuit.State) {
+	logger := log.NewLogEntry()
+
+	p.StatsdClient.Incr("provider.circuit_change", []string{fmt.Sprintf("from_state:%s", from), fmt.Sprintf("to_state:%s", to)}, 1.0)
+	logger.WithCircuitChangeFrom(from).WithCircuitChangeTo(to).Info("circuit breaker trigger")
+}
+
+// Redeem fulfills the Provider interface.
+// The authenticator uses this method to redeem the code provided to /callback after the user logs into their Okta account.
+// The code is redeemed for an access token and refresh token
+// 1. POSTs the code and grant_type to https://developer.okta.com/docs/api/resources/oidc/#token
+// 2. If the request fails, the authenticator will return a 500 and display an error page (see oauth_proxy.go#OAuthCallback)
+// 3. If the request succeeds, the data from Okta contains:
+//     - the access token which we use to get data from Okta
+//     - the refresh token which we can use to get a new access_token
+//     - the expiration time of the access token
+//     - a Base64 encoded id token which contains the user's email
+//       address and whether or not that email address is verified
+func (p *OktaProvider) Redeem(redirectURL, code string) (*sessions.SessionState, error) {
+	if code == "" {
+		return nil, ErrBadRequest
+	}
+	params := url.Values{}
+	params.Add("redirect_uri", redirectURL)
+	params.Add("client_id", p.ClientID)
+	params.Add("client_secret", p.ClientSecret)
+	params.Add("code", code)
+	params.Add("grant_type", "authorization_code")
+	params.Add("scope", p.Scope)
+
+	var response struct {
+		AccessToken  string `json:"access_token"`
+		RefreshToken string `json:"refresh_token"`
+		ExpiresIn    int64  `json:"expires_in"`
+		IDToken      string `json:"id_token"`
+	}
+	err := p.oktaRequest("POST", p.RedeemURL.String(), params, []string{"action", "redeem"}, nil, &response)
+	if err != nil {
+		return nil, err
+	}
+	var email string
+	email, err = p.verifyEmailWithAccessToken(response.AccessToken)
+	if err != nil {
+		return nil, err
+	}
+	return &sessions.SessionState{
+		AccessToken:  response.AccessToken,
+		RefreshToken: response.RefreshToken,
+
+		RefreshDeadline:  sessions.ExtendDeadline(time.Duration(response.ExpiresIn) * time.Second),
+		LifetimeDeadline: sessions.ExtendDeadline(p.SessionLifetimeTTL),
+		Email:            email,
+	}, nil
+}
+
+// verifyEmailWithAccessToken takes in an access token and
+// checks if it is verified with Okta. Conditionally returns
+// an empty string, or if validated the email address in question.
+func (p *OktaProvider) verifyEmailWithAccessToken(AccessToken string) (string, error) {
+	if AccessToken == "" {
+		return "", ErrBadRequest
+	}
+
+	userinfo, err := p.UserInfo(AccessToken)
+	if err != nil {
+		return "", err
+	}
+	if userinfo.EmailAddress == "" {
+		return "", errors.New("missing email")
+	}
+	if !userinfo.EmailVerified {
+		return "", fmt.Errorf("email %s not listed as verified", userinfo.EmailAddress)
```

Comment:
Let's not use a templated error struct here, but just return a generic `email not verified` and potentially log the error?

I think this will be nicer on structure log systems.